ambry-store/src/test/java/com.github.ambry.store/LogTest.java

```diff
@@ -270,6 +256,129 @@ public void getFileSpanForMessageBadArgsTest() throws IOException {
     }
   }
 
+  /**
+   * Tests all cases of {@link Log#addSegment(LogSegment, boolean)} and {@link Log#dropSegment(String, boolean)}.
+   * @throws IOException
+   */
+  @Test
+  public void addAndDropSegmentTest() throws IOException {
+    // start with a segment that has a high position to allow for addition of segments
+    long activeSegmentPos = 2 * LOG_CAPACITY / SEGMENT_CAPACITY;
+    LogSegment loadedSegment = getLogSegment(LogSegmentNameHelper.getName(activeSegmentPos, 0), SEGMENT_CAPACITY, true);
+    List<LogSegment> segmentsToLoad = Collections.singletonList(loadedSegment);
+    Log log = new Log(tempDir.getAbsolutePath(), LOG_CAPACITY, SEGMENT_CAPACITY, metrics, true, segmentsToLoad,
+        Collections.EMPTY_LIST.iterator());
+
+    // add a segment
+    String segmentName = LogSegmentNameHelper.getName(0, 0);
+    LogSegment uncountedSegment = getLogSegment(segmentName, SEGMENT_CAPACITY, true);
+    log.addSegment(uncountedSegment, false);
+    assertEquals("Log segment instance not as expected", uncountedSegment, log.getSegment(segmentName));
+
+    // cannot add past the active segment
+    segmentName = LogSegmentNameHelper.getName(activeSegmentPos + 1, 0);
+    LogSegment segment = getLogSegment(segmentName, SEGMENT_CAPACITY, true);
+    try {
+      log.addSegment(segment, false);
+      fail("Should not be able to add past the active segment");
+    } catch (IllegalArgumentException e) {
+      // expected. Nothing to do.
+    }
+
+    // since the previous one did not ask for count of segments to be increased, we should be able to add max - 1
+    // more segments. This increments used segment count to the max.
+    int max = (int) (LOG_CAPACITY / SEGMENT_CAPACITY);
+    for (int i = 1; i < max; i++) {
+      segmentName = LogSegmentNameHelper.getName(i, 0);
+      segment = getLogSegment(segmentName, SEGMENT_CAPACITY, true);
+      log.addSegment(segment, true);
+    }
```

Comment:
I think `addSegment()` should disallow addition of yet another segment and you should test that here.